subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c

```diff
@@ -224,6 +296,186 @@ static int validate_ak_order(void)
+	if (err) {
+		return err;
+	}
+
+	err = settings_delete(name);
+	if (err) {
+		return err;
+	}
+
+	return 0;
+}
+
+static int bond_delete(size_t bond_idx)
+{
+	int err;
+
+	err = fp_settings_bond_delete(bond_idx);
+	if (err) {
+		return err;
+	}
+
+	fp_bonds[bond_idx] = (struct fp_bond){0};
+
+	return 0;
+}
+
+static int bond_remove(size_t bond_idx)
+{
+	int err;
+
+	if (fp_bonds[bond_idx].bonded) {
+		/* Bond entry will be deleted in the fp_storage_ak_bond_deleted function called
+		 * after bond removal from Bluetooth stack.
+		 */
+		err = bt_request_cb->bond_remove(&fp_bonds[bond_idx].nv.addr);
+		if (err) {
+			LOG_ERR("Failed to remove bond (err %d)", err);
+			return err;
+		}
+	} else {
+		err = bond_delete(bond_idx);
+		if (err) {
+			LOG_ERR("Failed to delete bond (err %d)", err);
+			return err;
+		}
+	}
+
+	LOG_DBG("Bond removed successfully");
+
+	return 0;
+}
+
+static void bond_remove_old_id(void)
+{
+	for (size_t i = 0; i < ARRAY_SIZE(fp_bonds); i++) {
+		int err;
+		bool account_key_found = false;
+
+		if (bt_addr_le_eq(&fp_bonds[i].nv.addr, BT_ADDR_LE_ANY) || !fp_bonds[i].bonded ||
+		    (fp_bonds[i].nv.account_key_metadata_id == 0)) {
+			continue;
+		}
+
+		for (size_t j = 0; j < account_key_count; j++) {
+			if (fp_bonds[i].nv.account_key_metadata_id ==
+			    ACCOUNT_KEY_METADATA_FIELD_GET(account_key_metadata[j], ID)) {
+				account_key_found = true;
+				break;
+			}
+		}
+
+		if (!account_key_found) {
+			LOG_DBG("Removing bond associated with old Account Key");
+			err = bond_remove(i);
+			if (err) {
+				LOG_ERR("Failed to remove bond (err %d)", err);
+			}
+		}
+	}
+}
+
+static void bt_bonds_find(void)
+{
+	for (size_t i = 0; i < ARRAY_SIZE(fp_bonds); i++) {
+		if (!bt_request_cb->bond_find(&fp_bonds[i].nv.addr)) {
+			LOG_DBG("Loaded bond has been found in the Bluetooth bond list");
+			fp_bonds[i].bonded = true;
+		}
+	}
+}
+
+static void handle_bond_duplicates(void)
+{
+	for (size_t i = 0; i < ARRAY_SIZE(fp_bonds); i++) {
+		bool duplicate_found = false;
+		int err;
+
+		if (bt_addr_le_eq(&fp_bonds[i].nv.addr, BT_ADDR_LE_ANY)) {
+			continue;
+		}
+
+		for (size_t j = i + 1; j < ARRAY_SIZE(fp_bonds); j++) {
+			if (bt_addr_le_eq(&fp_bonds[j].nv.addr, &fp_bonds[i].nv.addr)) {
+				/* Error during Fast Pair Procedure. */
+				duplicate_found = true;
+				LOG_DBG("Removing both duplicated bond entries at bootup");
+				err = bond_delete(j);
+				if (err) {
+					LOG_ERR("Failed to delete bond (err %d)", err);
+				}
+			}
+		}
+
+		if (duplicate_found) {
+			err = bond_remove(i);
+			if (err) {
+				LOG_ERR("Failed to remove bond (err %d)", err);
+			}
+		}
+	}
+}
+
+static void validate_bonds(void)
+{
+	int err;
+
+	/* Assert that zero-initialized addresses are equal to BT_ADDR_LE_ANY. */
+	static const bt_addr_le_t addr;
+	__ASSERT_NO_MSG(bt_addr_le_eq(&addr, BT_ADDR_LE_ANY));
+
+	bt_bonds_find();
+
+	handle_bond_duplicates();
+
+	for (size_t i = 0; i < ARRAY_SIZE(fp_bonds); i++) {
```

Comment:
Perhaps also define a helper function for this code block. e.g `invalid_bonds_purge()`